File: policysquare/lib/data/models/action_item.dart
import 'package:json_annotation/json_annotation.dart';

part 'action_item.g.dart';

@JsonSerializable()
class ActionItem {
  String? id;
  String? visitId;
  String? propertyCode;
  String? location;
  String? category;
  String? issue;
  String? priority; // CRITICAL, HIGH, MEDIUM, LOW
  bool? safetyHazard;
  String? immediateControl;
  String? responsibleParty;
  String? tenantUnit;
  bool? leaseReview;
  String? actionOwner;
  String? status; // OPEN, QUOTE_REQUESTED, APPROVED, IN_PROGRESS, VERIFIED, CLOSED
  String? targetCompletion;
  String? completionDate;
  String? verifiedBy;
  double? estimatedCost;
  String? costClassification; // OPERATING, CAPITAL
  String? approvalStatus; // NOT_REQUIRED, PENDING, APPROVED, DECLINED
  String? vendorRef;
  String? evidenceLink;
  String? notes;
  String? mobileNumber;

  /// Derived server-side.
  bool? requiresCapitalReview;
  bool? open;
  bool? overdue;
  dynamic createdAt;

  ActionItem({
    this.id,
    this.visitId,
    this.propertyCode,
    this.location,
    this.category,
    this.issue,
    this.priority,
    this.safetyHazard,
    this.immediateControl,
    this.responsibleParty,
    this.tenantUnit,
    this.leaseReview,
    this.actionOwner,
    this.status,
    this.targetCompletion,
    this.completionDate,
    this.verifiedBy,
    this.estimatedCost,
    this.costClassification,
    this.approvalStatus,
    this.vendorRef,
    this.evidenceLink,
    this.notes,
    this.mobileNumber,
    this.requiresCapitalReview,
    this.open,
    this.overdue,
    this.createdAt,
  });

  factory ActionItem.fromJson(Map<String, dynamic> json) =>
      _$ActionItemFromJson(json);
  Map<String, dynamic> toJson() => _$ActionItemToJson(this);
}

File: backend/src/main/java/com/policysquare/commercial/model/ActionItem.java
package com.policysquare.commercial.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.policysquare.commercial.config.PmaSettings;
import jakarta.persistence.*;
import lombok.Data;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * A deficiency or follow-up item raised by an inspection visit, tracked from
 * discovery through quote, approval, completion and verification.
 */
@Entity
@Data
@Table(name = "action_items")
public class ActionItem {

    /** Stable action reference, e.g. "RR-20260912-01". */
    @Id
    private String id;

    @Column(name = "visit_id")
    private String visitId;

    @Column(name = "property_code", nullable = false)
    private String propertyCode;

    private String location;

    private String category;

    @Column(columnDefinition = "TEXT")
    private String issue;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private Priority priority;

    @Column(name = "safety_hazard")
    private Boolean safetyHazard = false;

    @Column(name = "immediate_control", columnDefinition = "TEXT")
    private String immediateControl;

    @Column(name = "responsible_party")
    private String responsibleParty;

    @Column(name = "tenant_unit")
    private String tenantUnit;

    @Column(name = "lease_review")
    private Boolean leaseReview = false;

    @Column(name = "action_owner")
    private String actionOwner;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private ActionStatus status;

    @Column(name = "target_completion")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private LocalDate targetCompletion;

    @Column(name = "completion_date")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private LocalDate completionDate;

    @Column(name = "verified_by")
    private String verifiedBy;

    @Column(name = "estimated_cost", precision = 15, scale = 2)
    private BigDecimal estimatedCost;

    @Column(name = "cost_classification")
    @Enumerated(EnumType.STRING)
    private CostClassification costClassification;

    @Column(name = "approval_status")
    @Enumerated(EnumType.STRING)
    private ApprovalStatus approvalStatus;

    @Column(name = "vendor_ref")
    private String vendorRef;

    @Column(name = "evidence_link")
    private String evidenceLink;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "mobile_number")
    private String mobileNumber;

    @Column(updatable = false)
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime createdAt = LocalDateTime.now();

    public LocalDateTime getCreatedAt() {
        return createdAt != null ? createdAt : LocalDateTime.now();
    }

    /** Work above the PMA threshold cannot proceed on operating approval alone. */
    @Transient
    public boolean isRequiresCapitalReview() {
        return estimatedCost != null
                && estimatedCost.compareTo(PmaSettings.CAPITAL_REVIEW_THRESHOLD) > 0;
    }

    @Transient
    public boolean isOpen() {
        return status != ActionStatus.CLOSED;
    }

    @Transient
    public boolean isOverdue() {
        return isOpen()
                && targetCompletion != null
                && LocalDate.now().isAfter(targetCompletion);
    }

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
    }

    public enum Priority {
        CRITICAL, HIGH, MEDIUM, LOW
    }

    public enum ActionStatus {
        OPEN, QUOTE_REQUESTED, APPROVED, IN_PROGRESS, VERIFIED, CLOSED
    }

    public enum CostClassification {
        OPERATING, CAPITAL
    }

    public enum ApprovalStatus {
        NOT_REQUIRED, PENDING, APPROVED, DECLINED
    }
}
